patches/api/0335-Add-methods-to-find-targets-for-lightning-strikes.patch

```diff
@@ -0,0 +1,47 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Jakub Zacek <dawon@dawon.eu>
+Date: Sun, 3 Oct 2021 13:08:06 +0200
+Subject: [PATCH] Add methods to find targets for lightning strikes
+
+
+diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
+index dc497606ebf9903c2d32276ecddda3231457346a..0221aa063a048e511821d1e6c985eef0fc61b943 100644
+--- a/src/main/java/org/bukkit/World.java
++++ b/src/main/java/org/bukkit/World.java
+@@ -757,6 +757,36 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
+     @NotNull
+     public LightningStrike strikeLightningEffect(@NotNull Location loc);
+ 
++    // Paper start
++    /**
++     * Finds closest Lightning Rod in 128 block radius around input location.
++     * If no such Lightning Rod is found, null will be returned.
++     *
++     * @param loc Location to search Lightning Rod around
++     * @return Location of Lightning Rod or null if not found
++     */
++    @Nullable
++    public Location findLightningRod(@NotNull Location loc);
++
++    /**
++     * Finds a target {@link Location} for lightning to strike to.
++     * <p>
++     * It looks for:
++     * <ol>
++     * <li>Lightning Rod using {@link World#findLightningRod(Location)}</li>
++     * <li>Random alive {@link LivingEntity} that can see sky
++     * in 10x10x10 cube around input {@link Location} </li>
++     * </ol>
++     *
++     * If neither target is found, input {@link Location} will be returned
```

Comment:
It will return a location with the same x and z as the input, but with the highest motion blocking pos as the y. Does it maybe make sense to return null instead, as we didn't find a "target" (either entity or lightning rod)?

Reply:
I totally agree with you on this one, but: How should I detect such a case? I see three possible imperfect solutions:

1a) comparing the resulting location's x&z against the source location. This solution risks the case when the target entity/rod was found on these x&z coords -> false negative. 
1b) I could check x&y&z (with finding highest motion blocking y) - I guess same as 1) applies but with less probability? (Motion blocking block == living entities can't go through it? - that would be also a lightning rod, right?
2) I could do the same as 1b) and then check that the resulting location is not a lightning rod. This still is risking an entity on these coordinates.
3) Modifying the original method (splitting it into 2 methods, where one returns null on non found). This should be fine, but is that acceptable by Paper standards? The smallest possible changes are encouraged and this would be quite a big one. Future maintenance of this would be hell...

Or am I overthinking it too much?

Reply:
Third way would be preferred, and the diff wouldn't really be big or difficult to maintain.

```diff
diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index 9e4ad810dd..cb697f382f 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -956,6 +956,11 @@ public class ServerLevel extends Level implements WorldGenLevel {
     }
 
     protected BlockPos findLightningTargetAround(BlockPos pos) {
+        // Paper start
+        return this.findLightningTargetAround(pos, false);
+    }
+    protected BlockPos findLightningTargetAround(BlockPos pos, boolean returnNullWhenNoTarget) {
+        // Paper end
         BlockPos blockposition1 = this.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING, pos);
         Optional<BlockPos> optional = this.findLightningRod(blockposition1);
 
@@ -970,6 +975,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
             if (!list.isEmpty()) {
                 return ((LivingEntity) list.get(this.random.nextInt(list.size()))).blockPosition();
             } else {
+                if (returnNullWhenNoTarget) return null; // Paper
                 if (blockposition1.getY() == this.getMinBuildHeight() - 1) {
                     blockposition1 = blockposition1.above(2);
                 }
```

Reply:
Ah ok, then. Will do, thank you!